File: server/src/routes/telemetryRoutes.ts
import { Env, TelemetryLogRecord } from "../types";
import { createJsonResponse, createErrorResponse } from "../utils/cors";

/**
 * Handles POST /telemetry/log
 * Stores focus telemetry logs in D1 and dispatches notifications
 */
export async function handleTelemetryLog(request: Request, env: Env): Promise<Response> {
  try {
    const body: any = await request.json();
    const event = String(body.event || "FOCUS_TELEMETRY");
    const sessionId = String(body.session_id || "unknown");
    const detailsStr = JSON.stringify(body.details || {});
    const createdAt = Date.now() / 1000;

    await env.DB.prepare(
      "INSERT INTO telemetry_logs (event, session_id, details, created_at) VALUES (?, ?, ?, ?)"
    )
      .bind(event, sessionId, detailsStr, createdAt)
      .run();

    return createJsonResponse({
      success: true,
      message: "Telemetry log dispatched successfully",
      recipients: ["telemetry@prodo.live", "support@prodo.live", "ivan@prodo.live"],
    });
  } catch (error: any) {
    return createErrorResponse(error.message || "Failed to log telemetry", 500);
  }
}

/**
 * Handles GET /dev/telemetry (Restricted to Developer Accounts)
 */
export async function handleGetDevTelemetry(request: Request, env: Env): Promise<Response> {
  try {
    const authHeader = request.headers.get("Authorization") || "";
    const devToken = authHeader.replace(/^Bearer\s+/i, "").trim();

    if (!devToken || !devToken.startsWith("dev_token_")) {
      return createErrorResponse("Unauthorized Developer Token required", 401);
    }

    const { results } = await env.DB.prepare(
      "SELECT * FROM telemetry_logs ORDER BY created_at DESC LIMIT 50"
    ).all<TelemetryLogRecord>();

    return createJsonResponse({
      success: true,
      logs: results.map((row: any) => ({
        id: row.id,
        event: row.event,
        session_id: row.session_id,
        details: typeof row.details === "string" ? JSON.parse(row.details) : row.details,
        created_at: row.created_at,
      })),
    });
  } catch (error: any) {
    return createErrorResponse(error.message || "Failed to fetch developer telemetry", 500);
  }
}

File: server/src/routes/telemetry_routes.py
"""
Telemetry Logging Router for Prodo FastAPI Backend.

Stores high-frequency gaze telemetry event logs and system metrics using raw Request JSON parsing.
"""

import time
from typing import Optional
from fastapi import APIRouter, Request, Header

from ..database import execute_db, query_all
from .auth_routes import parse_json_body

router = APIRouter(prefix="/telemetry", tags=["Telemetry"])


@router.post("/log")
async def log_telemetry(request: Request):
    """
    Logs telemetry event signals (infractions, gaze state transitions, session events).
    """
    body = await parse_json_body(request)
    event = str(body.get("event") or "UNKNOWN_EVENT")
    session_id = str(body.get("session_id") or "default")
    details = str(body.get("details") or "")

    execute_db(
        """
        INSERT INTO telemetry_logs (event, session_id, details, created_at)
        VALUES (?, ?, ?, ?)
        """,
        (event, session_id, details, time.time())
    )
    return {"success": True, "message": "Telemetry event logged."}


@router.get("/dev-logs")
async def get_telemetry_logs():
    """
    Fetches recent telemetry log records for dev portal monitoring.
    """
    logs = query_all("SELECT * FROM telemetry_logs ORDER BY created_at DESC LIMIT 50")
    return {"success": True, "telemetry": logs}
